### PR TITLE
Heatmap

### PR DESCRIPTION
**Pull request by ConstantinoSchillebeeckx**
This is a re-open of #2050 now that I've started working on it again. This PR isn't finished yet, but I wanted to get the ball rolling and get your input as I move forward.

Some of the things still to do:
- ~~unit test~~
- ~~documentation~~
- ~~example detailing all the chart options~~
- ~~various chart features~~
- ~~fix for [flashing cells](https://github.com/novus/nvd3/pull/2101/files#diff-cf90f721fe299a2b3f587de5fc06df72R6)~~

### CODE REVIEW

File: src/models/heatMapChart.js [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    // get max/min range for all the quantized cell values
+    // returns an array where each element is [start,stop]
+    // of color bin
+    function quantizeLegendValues() {
+
+        var e = heatMap.colorScale();
+
+        return e.range().map(function(color) {
+          var d = e.invertExtent(color);
+          if (d[0] == null) d[0] = e.domain()[0];
```
Comment (jeznag) [inline src/models/heatMapChart.js]: Any reason not to use === here?

File: src/models/heatMapChart.js [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        , showYAxis = true
+        , alignYAxis = 'left'
+        , alignXAxis = 'top'
+        , rotateLabels = 0
+        , title = false
+        , x
+        , y
+        , noData = null
+        , dispatch = d3.dispatch('beforeUpdate','renderEnd')
+        //, dispatch = d3.dispatch('beforeUpdate', 'elementMouseover', 'elementMouseout', 'elementMousemove', 'renderEnd')
```
Comment (jeznag) [inline src/models/heatMapChart.js]: can probably delete this

File: src/models/heatMapChart.js (hunk 1 above)
Comment (ConstantinoSchillebeeckx) [reply src/models/heatMapChart.js]: None at all, will fix.

**Comment by liquidpele** [comment]
Tried to load up your branch to view the example, but it throws javascript errors :(

**Comment by ConstantinoSchillebeeckx** [comment]
Could you show me a screenshot of the errors exactly? Works just fine on my end. 🙈

**Comment by liquidpele** [comment]
oh, it was just because your example tries to get d3 from bower_components instead of the cdn like the others... please change that because the examples get directly dumped to the site page at https://nvd3-community.github.io/nvd3/ which won't have bower_components there.

**Comment by liquidpele** [comment]
One comment is that on default, when I go between squares, there is a flash where it colors everything before then going back to coloring only the row/column.

**Comment by ConstantinoSchillebeeckx** [comment]
[yep](https://github.com/novus/nvd3/pull/2101/files#diff-cf90f721fe299a2b3f587de5fc06df72R6) that's caused by the empty space between the cells - I'll see what I can do for a fix.

**Comment by liquidpele** [comment]
Also in some cases it displays "NaN"... should maybe just be blank?

**Comment by liquidpele** [comment]
PS:  another cool option I've seen is where you can click the column/row headers and it keeps the row/column highlighted for you.    example:   http://rpubs.com/jcheng/mtcars-heatmap

**Comment by ConstantinoSchillebeeckx** [comment]
There's an option for choosing what to display in case of missing data: [`missingDataLabel`](https://github.com/novus/nvd3/pull/2101/files#diff-cf90f721fe299a2b3f587de5fc06df72R39).

I like the idea of clicking to lock the fade/highlight; I also like the dendrograms but I was waiting to implement those as a standalone chart and then potentially including it as part of this one.

**Comment by ConstantinoSchillebeeckx** [comment]
Unless people aren't happy with the example, this should be good to go.

I did have a question in general about handling "bad" data - if a cell value is provided twice (e.g. the row/col combination exists more than once in the data) there is no warning provided. Instead, the previous cell value is overwritten by the new.

What is the general approach regarding this type of behavior? Should I instead throw a `nv.utils.noData()`?

Thinking ahead, it may be convenient to have a global data parsing/checking function that ensures that are consistent.

**Comment by liquidpele** [comment]
I'd just do what your'e doing now, I think it's reasonable for that case.

**Comment by liquidpele** [comment]
Looks like you have conflicts with master too... need to merge nvd3/master back to your branch.

**Comment by ConstantinoSchillebeeckx** [comment]
WTH am I doing wrong here 😢

**Comment by ConstantinoSchillebeeckx** [comment]
what's going on with the unit tests? runs just fine locally.

**Comment by jeznag** [comment]
@ConstantinoSchillebeeckx RE unit tests: yeah it's weird. I created an issue a while ago: https://github.com/novus/nvd3/issues/2102

Not sure how Jenkins is configured. I tried playing around with connect timeouts but nothing worked.

**Comment by liquidpele** [comment]
@ConstantinoSchillebeeckx I merged this, but on building it now thows an error on your model:

````
  stack: 'Error\n    at new JS_Parse_Error (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:196:18)\n    at js_error (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:204:11)\n    at croak (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:680:9)\n    at token_error (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:688:9)\n    at unexpected (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:694:9)\n    at expr_atom (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1201:9)\n    at maybe_unary (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1363:19)\n    at expr_ops (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1398:24)\n    at maybe_conditional (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1403:20)\n    at maybe_assign (/Users/rpegues/git/nvd3/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1427:20)' }
>> Uglifying source build/nv.d3.js failed.
Warning: Uglification failed.
Unexpected token: operator (>). 
Line 7122 in build/nv.d3.js
```

It appears to be from this line, which is using ES6 notation which we haven't ever supported yet due to our continuing IE11 support...

https://github.com/ConstantinoSchillebeeckx/nvd3/blob/3ee77f6c3014fa2d801808973b9e8dad4652c2b5/src/models/heatMap.js#L375

If you can fix this quickly we can merge it in, but otherwise I'll have to revert the PR

**Pull request merged**